cryptography/src/bls12381/primitives/ops/aggregate.rs

```diff
@@ -0,0 +1,490 @@
+) -> Result<(), Error>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a V::Public>,
+    V::Public: 'a,
+{
+    // Aggregate public keys
+    //
+    // We can take advantage of the bilinearity property of pairings to aggregate public keys
+    // that have all signed the same message (as long as all public keys are unique).
+    let agg_public = combine_public_keys::<V, _>(public);
+
+    // Compute the hash of the message
+    let hm = namespace.map_or_else(
+        || hash_message::<V>(V::MESSAGE, message),
+        |ns| hash_message_namespace::<V>(V::MESSAGE, ns, message),
+    );
+
+    // Verify the signature
+    V::verify(agg_public.inner(), &hm, signature.inner())
+}
+
+/// Verifies an aggregate signature over multiple unique messages from a single public key.
+///
+/// Each entry is a tuple of (namespace, message). The signature must be the aggregate
+/// of all individual signatures.
+///
+/// # Warning
+///
+/// This function is vulnerable to signature malleability when used with signatures
+/// that were aggregated from different messages. An attacker can redistribute
+/// signature components between messages while keeping the aggregate unchanged.
+/// Use [`batch::verify_multiple_messages`](super::batch::verify_multiple_messages) instead when signatures are provided individually.
+///
+/// This function assumes a group check was already performed on `public` and `signature`.
+/// It is not safe to provide an aggregate public key or to provide duplicate messages.
+pub fn verify_multiple_messages<'a, V, I>(
+    public: &V::Public,
+    messages: I,
+    signature: &Signature<V>,
+    #[cfg_attr(not(feature = "std"), allow(unused_variables))] concurrency: usize,
+) -> Result<(), Error>
+where
+    V: Variant,
+    I: IntoIterator<Item = &'a (Option<&'a [u8]>, &'a [u8])> + Send + Sync,
+    I::IntoIter: Send + Sync,
+{
+    #[cfg(not(feature = "std"))]
+    let hm_sum = compute_hm_sum::<V, I>(messages);
+
+    #[cfg(feature = "std")]
+    let hm_sum = if concurrency == 1 {
+        compute_hm_sum::<V, I>(messages)
+    } else {
+        let pool = ThreadPoolBuilder::new()
+            .num_threads(concurrency)
+            .build()
+            .expect("Unable to build thread pool");
+
+        pool.install(move || {
+            messages
+                .into_iter()
+                .par_bridge()
+                .map(|(namespace, msg)| {
+                    namespace.as_ref().map_or_else(
+                        || hash_message::<V>(V::MESSAGE, msg),
+                        |namespace| hash_message_namespace::<V>(V::MESSAGE, namespace, msg),
+                    )
+                })
+                .reduce(V::Signature::zero, |mut sum, hm| {
+                    sum += &hm;
+                    sum
+                })
+        })
+    };
+
+    V::verify(public, &hm_sum, signature.inner())
+}
+
+/// Computes the sum over the hash of each message.
+fn compute_hm_sum<'a, V, I>(messages: I) -> V::Signature
```

Comment:
I think it would be more appropriate to create a utility for hashing messages, and just do away with this method.